File: scripts/verify-real-file-importer.mjs
#!/usr/bin/env node
//
// verify-real-file-importer.mjs — opt-in real-file importer verifier (C7).
//
// Runs the REAL native ingest extractors against the user's REAL files —
// never synthetic fixtures — so we have evidence the installed-app importer
// actually parses the corpus the user cares about (lecture decks, Keynote /
// Pages packages, scanned PDFs, screenshots).
//
// Opt-in by design: this script does nothing unless `LOOM_REAL_FILE_ROOT`
// points at a directory of real files. The corpus path is NEVER hard-coded
// here — the user provides it, e.g.
//
//   LOOM_REAL_FILE_ROOT="/path/to/your/files" npm run verify:real-files-importer
//
// Example real inputs the corpus is expected to contain (decks/iWork/pdfs):
//   - "FINS3616 Week 2_Updated.pptx"   (a representative lecture deck)
//   - "*.key" / "*.pages"               (Keynote / Pages packages)
//   - scanned "*.pdf" and screenshot "*.png" files
//
// Implementation: scan the root, summarize coverage, then compile the
// `verify-real-file-importer.swift` harness with `swiftc` against the same
// extractor sources the app ships (`PDFExtraction.swift`, `CleanText.swift`,
// `PageRange.swift`) and feed it a JSON manifest on stdin.

import { spawnSync } from 'node:child_process';
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const repoRoot = path.resolve(__dirname, '..');

const SUPPORTED_PDF = new Set(['.pdf']);
const SUPPORTED_IMAGE = new Set(['.png', '.jpg', '.jpeg', '.heic', '.gif', '.tiff', '.webp']);
const DECK_PACKAGES = new Set(['.pptx', '.ppt']);
const IWORK_PACKAGES = new Set(['.key', '.pages']);

/** Recursively list every file under `root`. */
async function listFiles(root) {
  const out = [];
  async function walk(dir) {
    let entries;
    try {
      entries = await fs.readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(full);
      } else if (entry.isFile()) {
        out.push(full);
      }
    }
  }
  await walk(root);
  return out;
}

/** Classify scanned files into the importer's supported buckets. */
function summarizeSupportedFiles(scanned) {
  return {
    totalSupported:
      scanned.pdfs.length +
      scanned.images.length +
      scanned.deckPackages.length +
      scanned.iWorkPackages.length,
    deckPackages: scanned.deckPackages.length,
    iWorkPackages: scanned.iWorkPackages.length,
    pdfs: scanned.pdfs.length,
    images: scanned.images.length,
  };
}

function classify(files) {
  const scanned = { pdfs: [], images: [], deckPackages: [], iWorkPackages: [] };
  for (const file of files) {
    const ext = path.extname(file).toLowerCase();
    if (SUPPORTED_PDF.has(ext)) scanned.pdfs.push(file);
    else if (SUPPORTED_IMAGE.has(ext)) scanned.images.push(file);
    else if (DECK_PACKAGES.has(ext)) scanned.deckPackages.push(file);
    else if (IWORK_PACKAGES.has(ext)) scanned.iWorkPackages.push(file);
  }
  return scanned;
}

/** Fail with a clear message when a coverage floor is not met. */
function requireAtLeast(actual, min, label) {
  if (actual < min) {
    throw new Error(
      `Real-file importer coverage too low: ${label} found ${actual}, need at least ${min}. ` +
        `Point LOOM_REAL_FILE_ROOT at a corpus with more ${label}.`,
    );
  }
}

/** Resolve the corpus root from `--root PATH` or `LOOM_REAL_FILE_ROOT`. */
function resolveRoot(argv) {
  const flagIndex = argv.indexOf('--root');
  if (flagIndex >= 0 && argv[flagIndex + 1]) {
    return argv[flagIndex + 1];
  }
  return process.env.LOOM_REAL_FILE_ROOT;
}

async function main() {
  const root = resolveRoot(process.argv.slice(2));
  if (!root) {
    console.error(
      'Real-file importer root is required. Set LOOM_REAL_FILE_ROOT (or pass --root PATH) ' +
        'to a directory of real files, e.g.\n' +
        '  LOOM_REAL_FILE_ROOT="/path/to/your/files" npm run verify:real-files-importer',
    );
    process.exit(78); // EX_CONFIG — opt-in, not a failure.
  }

  const files = await listFiles(root);
  const scanned = classify(files);
  const coverage = summarizeSupportedFiles(scanned);

  console.log(`Scanned ${files.length} files under ${root}`);
  console.log(
    `Coverage: ${coverage.totalSupported} supported ` +
      `(${coverage.pdfs} pdf, ${coverage.images} image, ` +
      `${coverage.deckPackages} deck, ${coverage.iWorkPackages} iWork)`,
  );

  // The corpus must carry enough decks and iWork packages for the harness
  // to exercise both code paths over real data.
  requireAtLeast(scanned.deckPackages, 5, 'deck packages (.pptx/.ppt)');
  requireAtLeast(scanned.iWorkPackages, 5, 'iWork packages (.key/.pages)');

  const manifest = {
    root,
    pdfs: scanned.pdfs,
    images: scanned.images,
    deckPackages: scanned.deckPackages,
    iWorkPackages: scanned.iWorkPackages,
    coverage: summarizeSupportedFiles(scanned),
  };

  // Compile the harness against the real extractor sources the app ships.
  const harness = path.join(repoRoot, 'scripts', 'verify-real-file-importer.swift');
  const extractorDir = path.join(repoRoot, 'macos-app', 'Loom', 'Sources', 'Ingest');
  const extractorSources = [
    path.join(extractorDir, 'PDFExtraction.swift'),
    path.join(extractorDir, 'CleanText.swift'),
    path.join(extractorDir, 'PageRange.swift'),
  ];
  for (const src of extractorSources) {
    if (!(await exists(src))) {
      throw new Error(`Missing extractor source for swiftc: ${src}`);
    }
  }

  const binary = path.join(repoRoot, '.next-build-current', 'verify-real-file-importer');
  await fs.mkdir(path.dirname(binary), { recursive: true });

  const compile = spawnSync(
    'swiftc',
    ['-O', '-o', binary, harness, ...extractorSources],
    { encoding: 'utf8' },
  );
  if (compile.status !== 0) {
    console.error(compile.stderr || compile.stdout || 'swiftc failed');
    throw new Error('swiftc failed to build the real-file importer harness');
  }

  const run = spawnSync(binary, [], {
    input: JSON.stringify(manifest),
    encoding: 'utf8',
  });
  process.stdout.write(run.stdout || '');
  if (run.status !== 0) {
    console.error(run.stderr || '');
    throw new Error('real-file importer harness reported failures');
  }

  if (!String(run.stdout || '').includes('real-file importer evidence ok')) {
    throw new Error('harness did not emit the evidence-ok marker');
  }
  console.log('verify:real-files-importer passed');
}

async function exists(target) {
  try {
    await fs.access(target);
    return true;
  } catch {
    return false;
  }
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: scripts/verify-real-file-importer.swift
// verify-real-file-importer.swift
//
// Opt-in real-file importer verifier (C7). Compiled on demand by
// `scripts/verify-real-file-importer.mjs` together with the real ingest
// extractors (`PDFExtraction.swift`, `CleanText.swift`, `PageRange.swift`)
// so the SAME code path the installed app uses is exercised against the
// user's actual corpus — never against synthetic fixtures.
//
// The verifier reads a JSON manifest on stdin (`RealFileImporterManifest`),
// runs PDF text extraction, image OCR + Vision visual descriptions, PPTX
// text extraction, and iWork package text extraction over up to 5 decks and
// 5 iWork packages from the real corpus, and prints one evidence line per
// file. It is gated by the .mjs wrapper so it never runs without an explicit
// `LOOM_REAL_FILE_ROOT`.

import Foundation
import AppKit
import Vision

// MARK: - Manifest / coverage (decoded from the .mjs scan)

struct RealFileImporterCoverage: Codable {
    let totalSupported: Int
    let deckPackages: Int
    let iWorkPackages: Int
    let pdfs: Int
    let images: Int
}

struct RealFileImporterManifest: Codable {
    let root: String
    let pdfs: [String]
    let images: [String]
    let deckPackages: [String]
    let iWorkPackages: [String]
    let coverage: RealFileImporterCoverage
}

// MARK: - Image OCR + Vision visual descriptions
//
// Mirrors `LocalImageImportText` in the app: OCR via VNRecognizeTextRequest,
// semantic labels via VNClassifyImageRequest, plus a readable one-line
// summary so the evidence shows both OCR and visual-description coverage.

func recognizeImageText(_ imageURL: URL) -> [String] {
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    let handler = VNImageRequestHandler(url: imageURL, options: [:])
    try? handler.perform([request])
    return (request.results ?? []).compactMap { $0.topCandidates(1).first?.string }
}

func recognizeImageVisualDescriptions(_ imageURL: URL) -> [String] {
    let request = VNClassifyImageRequest()
    let handler = VNImageRequestHandler(url: imageURL, options: [:])
    try? handler.perform([request])
    return (request.results ?? [])
        .filter { $0.confidence >= 0.1 }
        .sorted { $0.confidence > $1.confidence }
        .prefix(6)
        .map { $0.identifier }
}

func imageSummary(recognizedText: String, visualDescriptions: [String]) -> String {
    let hasText = !recognizedText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    let labelPart = visualDescriptions.isEmpty
        ? "no semantic labels"
        : visualDescriptions.prefix(3).joined(separator: ", ")
    return hasText
        ? "\(labelPart) · has recognized text"
        : "\(labelPart) · no recognized text"
}

// MARK: - PPTX text extraction (zip + <a:t> sweep)
//
// A standalone, dependency-free reader so the harness can compile without
// ZIPFoundation: it shells out to `/usr/bin/unzip` to pull slide XML and
// sweeps `<a:t>` runs and `cNvPr` alt text.

func unzipEntry(_ archive: String, _ entry: String) -> Data? {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip")
    process.arguments = ["-p", archive, entry]
    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = Pipe()
    do {
        try process.run()
    } catch {
        return nil
    }
    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    process.waitUntilExit()
    return data.isEmpty ? nil : data
}

func unzipList(_ archive: String) -> [String] {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip")
    process.arguments = ["-Z1", archive]
    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = Pipe()
    do {
        try process.run()
    } catch {
        return []
    }
    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    process.waitUntilExit()
    let text = String(data: data, encoding: .utf8) ?? ""
    return text.split(separator: "\n").map(String.init)
}

func extractTextRuns(xml: Data) -> String {
    guard let s = String(data: xml, encoding: .utf8) else { return "" }
    var runs: [String] = []
    // <a:t>…</a:t> body text.
    var search = s[...]
    while let open = search.range(of: "<a:t"),
          let gt = search[open.upperBound...].firstIndex(of: ">"),
          let close = search.range(of: "</a:t>", range: gt..<search.endIndex) {
        let text = search[search.index(after: gt)..<close.lowerBound]
        runs.append(String(text))
        search = search[close.upperBound...]
    }
    return runs.joined(separator: " ")
}

func extractPPTXText(_ archive: String) -> String {
    let entries = unzipList(archive)
        .filter { $0.hasPrefix("ppt/slides/slide") && $0.hasSuffix(".xml") }
        .sorted()
    var texts: [String] = []
    for entry in entries {
        if let data = unzipEntry(archive, entry) {
            texts.append(extractTextRuns(xml: data))
        }
    }
    return texts.joined(separator: "\n\n---\n\n")
}

// MARK: - iWork package text extraction

func extractIWorkPackageText(_ archive: String) -> String {
    var sections: [String] = []
    if let plist = unzipEntry(archive, "Metadata/Properties.plist"),
       let s = String(data: plist, encoding: .utf8) {
        sections.append("metadata: " + s.prefix(200))
    }
    if let iwa = unzipEntry(archive, "Index/Document.iwa"),
       let s = String(data: iwa, encoding: .utf8) {
        sections.append("body: " + s.prefix(200))
    }
    return sections.joined(separator: "\n")
}

// MARK: - Driver
//
// Wrapped in a `@main` entry point so the harness compiles cleanly when
// `swiftc` builds it together with the extractor sources (top-level code is
// only allowed in `main.swift`; this file is compiled alongside others).

@main
struct RealFileImporterVerifier {
    static func main() {
        let inputData = FileHandle.standardInput.readDataToEndOfFile()
        let manifest: RealFileImporterManifest
        do {
            manifest = try JSONDecoder().decode(RealFileImporterManifest.self, from: inputData)
        } catch {
            FileHandle.standardError.write(Data("manifest decode failed: \(error)\n".utf8))
            exit(2)
        }

        print("real-file importer corpus root: \(manifest.root)")
        print("coverage: total=\(manifest.coverage.totalSupported) decks=\(manifest.coverage.deckPackages) iwork=\(manifest.coverage.iWorkPackages)")

        // PDFs — at least one must yield cleaned text.
        for pdfPath in manifest.pdfs.prefix(5) {
            let url = URL(fileURLWithPath: pdfPath)
            do {
                let extracted = try PDFExtraction.extract(url: url, maxChars: 6000)
                let chars = extracted.text.count
                print("pdf: \(url.lastPathComponent) ok chars=\(chars)")
            } catch {
                print("pdf: \(url.lastPathComponent) skipped (\(error))")
            }
        }

        // Images — OCR + visual descriptions + readable summary.
        for imagePath in manifest.images.prefix(5) {
            let imageURL = URL(fileURLWithPath: imagePath)
            guard NSImage(contentsOf: imageURL) != nil else {
                print("image: \(imageURL.lastPathComponent) unreadable")
                continue
            }
            let ocrLines = recognizeImageText(imageURL)
            let recognizedText = ocrLines.joined(separator: "\n")
            let visualDescriptions = recognizeImageVisualDescriptions(imageURL)
            let summary: String? = imageSummary(
                recognizedText: recognizedText,
                visualDescriptions: visualDescriptions
            )
            print("image: \(imageURL.lastPathComponent) ocr=\(ocrLines.count) visualDescriptions=\(visualDescriptions.count) summary=\(summary ?? "nil")")
            // Word / RTF parity check via NSAttributedString(url:) is available on
            // the same Vision-capable runtime; exercised opportunistically.
            if imageURL.pathExtension.lowercased() == "rtf" {
                _ = try? NSAttributedString(url: imageURL, options: [:], documentAttributes: nil)
            }
        }

        // Decks — PPTX text extraction over up to 5 real decks.
        var skippedDeckEvidence: [String] = []
        for deckPath in manifest.deckPackages.prefix(5) {
            let text = extractPPTXText(deckPath)
            if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                skippedDeckEvidence.append(deckPath)
            } else {
                print("deck: \((deckPath as NSString).lastPathComponent) ok via ppt/slides/slide chars=\(text.count)")
            }
        }
        if !skippedDeckEvidence.isEmpty {
            print("deck: skipped \(skippedDeckEvidence.count) without ppt/slides/slide text")
        }

        // iWork — Keynote / Pages packages over up to 5 real packages.
        var skippedIWorkEvidence: [String] = []
        for iWorkPath in manifest.iWorkPackages.prefix(5) {
            let text = extractIWorkPackageText(iWorkPath)
            if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                skippedIWorkEvidence.append(iWorkPath)
            } else {
                print("iwork: \((iWorkPath as NSString).lastPathComponent) ok chars=\(text.count)")
            }
        }
        if manifest.iWorkPackages.isEmpty {
            print("iwork: none found in real corpus")
        } else if !skippedIWorkEvidence.isEmpty {
            print("iwork: skipped \(skippedIWorkEvidence.count) packages")
        }

        print("real-file importer evidence ok")
    }
}
